crates/bifrost/src/providers/replicated_loglet/network.rs

```diff
@@ -58,10 +63,164 @@ impl RequestPump {
+            "handle-append-request",
+            None,
+            task.run(incoming),
+        );
+    }
+}
+
+struct HandleAppendTask<T> {
+    provider: Arc<ReplicatedLogletProvider<T>>,
+    metadata: Metadata,
+}
+
+impl<T> HandleAppendTask<T>
+where
+    T: TransportConnect,
+{
+    fn get_loglet(
+        &self,
+        provider: &ReplicatedLogletProvider<T>,
+        header: &CommonRequestHeader,
+    ) -> Result<Arc<ReplicatedLoglet<T>>, SequencerStatus> {
+        let logs = self.metadata.logs();
+        let chain = logs
+            .chain(&header.log_id)
+            .ok_or(SequencerStatus::BadLogletId)?;
+
+        // appends are only possible to tail segment.
+        let segment = chain
+            .iter()
+            .rev()
+            .find(|segment| segment.index() == header.segment_index)
+            .ok_or(SequencerStatus::BadLogletId)?;
+
+        let loglet = provider
+            .get_replicated_loglet(header.log_id, header.segment_index, &segment.config.params)
+            .map_err(|err| SequencerStatus::Error(err.to_string()))?;
+
+        Ok(loglet)
+    }
+
+    async fn run(self, request: Incoming<Append>) -> anyhow::Result<()> {
+        let (reciprocal, append) = request.split();
+        let cancelled = cancellation_watcher();
+
+        let result = tokio::select! {
+            _ = cancelled => {
+                // we could try to respond to the request still
+                // but it's probably too late.
+                anyhow::bail!(ShutdownError);
+            },
+            result = self.append(&self.provider, append) => {
+                result
+            }
+        };
+
+        let appended = match result {
+            Ok(appended) => appended,
+            Err(err) => Appended {
+                first_offset: LogletOffset::INVALID,
+                header: CommonResponseHeader {
+                    known_global_tail: None,
+                    sealed: None,
+                    status: err,
+                },
+            },
+        };
+
+        reciprocal.prepare(appended).send().await?;
         Ok(())
     }
+
+    async fn append(
+        &self,
+        provider: &ReplicatedLogletProvider<T>,
+        append: Append,
+    ) -> Result<Appended, SequencerStatus> {
+        let loglet = self.get_loglet(provider, &append.header)?;
+
+        // We get the sequencer instead of calling loglet.enqueue_batch to make sure this
+        // is the sequencer node. Otherwise, the client gets an error and should retry against
+        // the sequencer node.
+        if !loglet.is_sequencer_local() {
+            return Err(SequencerStatus::NotSequencer);
+        }
+
+        let global_tail = loglet.known_global_tail();
+        // This is okay to be done in scope, since it's the only way to exert some back pressure on the callers
+        // until the sequencer has enough capacity to accept these records.
+        // This call should return immediately once records are scheduled for commit.
```

Comment:
Comment is out of sync with what's happening. We are performing the append in a task so we are not putting any back-pressure on the network channel.